src/expr.rs

```diff
@@ -51,14 +51,41 @@ pub enum OpLevel {
 ///
 /// This is separate from ExprImpl because we want expressions to be wrapped in an Rc,
 /// and we can't directly implement std::ops::Add, etc., for Rc&lt;E: ExprImpl&lt;Tgt;&gt;.
-#[derive(Debug,Clone)]
+#[derive(Debug, Clone)]
 pub struct Expr<T: TensorType> {
     expr: Rc<ExprImpl<T>>,
 }
 
 impl<T: TensorType> Expr<T> {
-    /// Returns the derivative of the expression with respect to the given variable.
-    pub fn derivative_by_variable(&self, var: &str) -> Result<Expr<T>, Status> {
+    pub fn new<I>(expr: I) -> Expr<T>
+    where
+        I: ExprImpl<T> + 'static,
+    {
+        Expr {
+            expr: Rc::new(expr),
+        }
+    }
+}
+
+impl<T: TensorType> ExprImpl<T> for Expr<T> {
```

Comment:
I don't think Expr should implement ExprImpl, because that would allow users to do things like `Expr::new(Expr::new(something))`, which is essentially `Rc<Rc<SomeImpl>>`.  If you're trying to expose the underlying ExprImpl, the implementing Deref for Expr is probably a better approach.

Reply:
I've never implemented Deref before, but it seems like exactly what I need yeah.